src/model/balance.js

```diff
@@ -7,64 +7,79 @@ const errors = require('../util/errors')
 const NotAcceptedError = errors.NotAcceptedError
 const InvalidFieldsError = errors.InvalidFieldsError
 
+const BALANCE_PREFIX = 'balance__'
+
 module.exports = class Balance extends EventEmitter {
   constructor (opts) {
     super()
 
     this._maximum = new BigNumber(opts.maximum)
+    this._balance = null
+
+    this._key = BALANCE_PREFIX
     this._store = opts.store
-    this._cached = null
 
-    // all reserved DB key prefixes are 9 characters
-    // so it lines up with 'transfer_'
-    this._key = 'balance__'
+    // used to keep writes to store in order. See queueWrite.
+    this._write = Promise.resolve()
   }
 
-  * get () {
-    return (yield this._getNumber()).toString()
+  // _queueWrite solves the problem of balance updates being committed to
+  // the store out of order. Because calling balance gets and puts asynchronously
+  // would cause a race condition, we keep a promise chain in `this._write`
+  // and attach each balance update to the end via `.then()`. No update to the
+  // balance will be committed to the database unless all previous writes are
+  // resolved.
+  _queueWrite () {
```

Comment:
If this is only used to write the balance to the store the function should have a more descriptive name